syntax = "proto3";

service Game {
  rpc Roll(RollRequest) returns (RollResponse);
  rpc Reroll(RerollRequest) returns (RerollResponse);
  rpc Write(WriteRequest) returns (WriteResponse);
  rpc GetGameStatus(GetGameStatusRequest) returns (GetGameStatusResponse);
};

message RollRequest {}

message RollResponse {
  repeated int32 dices = 1;
  int32 reroll_times = 2;
  GameStatus status = 3;
}

message RerollRequest {
  string session = 1;
  repeated int32 kept_indexes = 2;
}

message RerollResponse {
  string session = 1;
  repeated int32 dices = 2;
  int32 reroll_times = 3;
  GameStatus status = 4;
}

message WriteRequest {
  string session = 1;
  ScoreType type = 2;
}

message WriteResponse {
  int32 score = 1;
  GameStatus status = 2;
}

message GetGameStatusRequest { string session = 1; }

message GetGameStatusResponse {
  string session = 1;
  GameStatus status = 2;
}

enum ScoreType {
  SCORETYPE_INVALID = 0;
  Ones = 1;
  Twos = 2;
  Threes = 3;
  Fours = 4;
  Fives = 5;
  Sixes = 6;
  Chance = 7;
  FourOfAKind = 8;
  FullHouse = 9;
  SmallStraight = 10;
  LargeStraight = 11;
  Yahtzee = 12;
  Bonus = 13;
}

enum Action {
  GameOver = 0;
  Roll = 1;
  Reroll = 2;
  Write = 3;
}

message PartScore {
  ScoreType type = 1;
  int32 score = 2;
}

message GameStatus {
  repeated PartScore scores = 1;
  repeated int32 dices = 2;
  int32 reroll_times = 3;
  repeated Action available_actions = 4;
  int32 score = 5;
  int32 turns = 6;
}